codebase2/codebase/U/Codebase/Branch.hs

```diff
@@ -19,16 +35,23 @@ type MetadataValue = Reference
 
 data MdValues = MdValues (Map MetadataValue MetadataType) deriving (Eq, Ord, Show)
 
-type Causal m = C.Causal m CausalHash BranchHash (Branch m)
+type CausalBranch m = Causal m CausalHash BranchHash (Branch m)
 
--- | V2.Branch is like V1.Branch0; I would rename it, at least temporarily, but too hard.
+-- | A re-imagining of Unison.Codebase.Branch which is less eager in what it loads,
+-- which can often speed up load times and keep fewer things in memory.
 data Branch m = Branch
   { terms :: Map NameSegment (Map Referent (m MdValues)),
     types :: Map NameSegment (Map Reference (m MdValues)),
     patches :: Map NameSegment (PatchHash, m Patch),
-    children :: Map NameSegment (Causal m)
+    children :: Map NameSegment (CausalBranch m)
   }
 
+instance AsEmpty (Branch m) where
```

Comment:
Do we use this? I think just having separate top-level `null` / `empty` symbols might be simpler than a prism that can technically do both.

Reply:
Implementing this typeclass gives us the `Empty` pattern, which I've found pretty helpful in general. We definitely don't "need" it, but I do find the `Empty` pattern more readable in general than something like:

```haskell
\case
  b | null b -> ...
    | otherwise -> ...
```

or w/e.

I don't plan to use the prism itself, but the pattern is implemented via the prism.

Reply:
Oh, I see.

Hrm, well that still seems worse to me, at first blush 😅. What kind of stuff would you use the `Empty` pattern for?

I'm imagining code like:

```haskell
case b of
  Empty -> ...
  _ -> ...
```

which would leave me wondering, "so Empty is a data constructor?"

versus just

```haskell
if null b
  then ...
  else ...
```

Reply:
I suppose if you have a nested pattern match it might be useful.